File: 652. Find Duplicate Subtrees/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        self.hash, self.seen, self.res = dict(), dict(), []
        self.hashing(root)
        return self.res
    def hashing(self, curr):
        s = ["*C", str(curr.val)]
        if curr.left or curr.right:
            if curr.left and curr.left not in self.hash: self.hashing(curr.left)
            if curr.right and curr.right not in self.hash: self.hashing(curr.right)
            s.append(self.hash[curr.left] if curr.left else "*L")
            s.append(self.hash[curr.right] if curr.right else "*L")
        p = "".join(s)
        if p in self.seen and self.seen[p] == 1: self.res.append(curr)
        self.seen[p] = 1 if p not in self.seen else self.seen[p] + 1
        self.hash[curr] = p

File: 652. Find Duplicate Subtrees/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<TreeNode, String> hash;
    Map<String, Integer> seen;
    List<TreeNode> res;
    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        hash = new HashMap<>();
        seen = new HashMap<>();
        res = new ArrayList<>();
        hashing(root);
        return res;
    }
    public void hashing(TreeNode curr) {
        StringBuilder str = new StringBuilder();
        str.append("*C");
        str.append(curr.val);
        if (curr.left != null || curr.right != null) {
            if (curr.left != null && !hash.containsKey(curr.left)) hashing(curr.left);
            if (curr.right != null && !hash.containsKey(curr.right)) hashing(curr.right);
            str.append(curr.left != null ? hash.get(curr.left) : "*L");
            str.append(curr.right != null ? hash.get(curr.right) : "*R");
        }
        String s = str.toString();
        if (seen.getOrDefault(s, 0) == 1) res.add(curr);
        seen.put(s, seen.getOrDefault(s, 0) + 1);
        hash.put(curr, s);
    }
}